aio.core/aio/core/directory/abstract/directory.py

```diff
@@ -0,0 +1,225 @@
+
+
+class ADirectory(metaclass=abstracts.Abstraction):
+    """A filesystem directory, with an associated fileset, and tools for
+    finding files.
+
+    The required `path` argument should be the path to an existing filesystem
+    directory.
+
+    The `exclude` and `exclude_dirs` parameters are passed to the `grep`
+    command as `--exclude` and `--exclude-dir` args respectively.
+
+    You can specify inclusive and exclusive regex matchers with `path_matcher`
+    and `exclude_matcher`.
+
+    By default, only text files (from grep/git grep pov) will be searched,
+    you can override this by setting `text_only` to `False`.
+    """
+
+    def __init__(
+            self,
+            path: Union[str, pathlib.Path],
+            exclude: Optional[Iterable[str]] = None,
+            exclude_dirs: Optional[Iterable[str]] = None,
+            path_matcher: Optional[Pattern[str]] = None,
+            exclude_matcher: Optional[Pattern[str]] = None,
+            text_only: Optional[bool] = True) -> None:
+        self._path = path
+        self.path_matcher = path_matcher
+        self.exclude_matcher = exclude_matcher
+        self.text_only = text_only
+        self.exclude = exclude or ()
+        self.exclude_dirs = exclude_dirs or ()
+
+    @async_property(cache=True)
+    async def files(self) -> Set[str]:
+        """Set of relative file paths associated with this directory."""
+        return await self.grep(["-l", ""])
+
+    @cached_property
+    def grep_args(self) -> Tuple[str, ...]:
+        """Default args to pass when calling `grep`."""
+        return (
+            *self.grep_command_args,
+            *(("-I", )
+              if self.text_only
+              else ()),
+            *self.grep_exclusion_args)
+
+    @property
+    def grep_command_args(self) -> Tuple[str, ...]:
+        """Path args for the `grep` command."""
+        grep_command = shutil.which("grep")
+        if grep_command:
+            return grep_command, "-r"
+        raise exceptions.DirectoryError("Unable to find `grep` command")
+
+    @property
+    def grep_exclusion_args(self) -> Tuple[str, ...]:
+        """Grep flags to exclude paths and directories."""
+        return (
+            tuple(
+                f"--exclude={exclusion}"
+                for exclusion
+                in self.exclude)
+            + tuple(
+                f"--exclude-dir={directory}"
+                for directory
+                in self.exclude_dirs))
+
+    @cached_property
+    def path(self) -> pathlib.Path:
+        """Path to this directory."""
+        return pathlib.Path(self._path)
+
+    @cached_property
+    def shell(self) -> subprocess.AAsyncShell:
+        """Shell that uses the directory path as `cwd`."""
+        return subprocess.AsyncShell(cwd=self.path)
+
+    @lru_cache(maxsize=None)
+    def absolute_path(self, path: str) -> str:
+        """Make a directory-relative path absolute."""
+        return str(self.path.joinpath(path))
```

Comment:
can't use Path.resolve() from pathlib?

Reply:
yep, it probably should, i think it works already because `self.path` is already absolute but i guess there is no guarantee of that and it should resolve the full path - thanks, ill update...